src/openlcb/IfTcpImpl.hxx

```diff
@@ -0,0 +1,553 @@
+    /// @param src the rendered TCP message.
+    /// @param tgt the output generic message
+    /// @return false if the message is not well formatted or
+    /// it is not an OpenLCB message.
+    static bool parse_tcp_message(const string &src, GenMessage *tgt)
+    {
+        int expected_size = get_tcp_message_len(src.data(), src.size());
+        if (expected_size > (int)src.size() || expected_size < MIN_MESSAGE_SIZE)
+        {
+            LOG(WARNING, "Incomplete or incorrectly formatted TCP message.");
+            return false;
+        }
+        uint16_t flags = data_to_error(&src[0]);
+        if ((flags & FLAGS_OPENLCB_MSG) == 0)
+        {
+            return false;
+        }
+        HASSERT((flags & FLAGS_CHAINING) == 0);
+        tgt->flagsDst = 0;
+        tgt->flagsSrc = 0;
+        if (flags & FLAGS_FRAGMENT_NOT_FIRST)
+        {
+            tgt->set_flag_dst(GenMessage::DSTFLAG_NOT_FIRST_MESSAGE);
+        }
+        if (flags & FLAGS_FRAGMENT_NOT_LAST)
+        {
+            tgt->set_flag_dst(GenMessage::DSTFLAG_NOT_LAST_MESSAGE);
+        }
+        // Contents of the size field in the header.
+        uint32_t sz = (uint8_t)src[HDR_SIZE_OFS];
+        sz <<= 8;
+        sz |= (uint8_t)src[HDR_SIZE_OFS + 1];
+        sz <<= 8;
+        sz |= (uint8_t)src[HDR_SIZE_OFS + 2];
+        if ((sz + HDR_SIZE_OFS) > src.size())
+        {
+            return false;
+        }
+        const char *msg = &src[HDR_LEN];
+        // We have already checked the size to be long enough for a global
+        // message with 0 bytes payload. So source address and MTI is ok to
+        // parse now.
+        tgt->mti = (Defs::MTI)data_to_error(msg + MSG_MTI_OFS);
+        tgt->src.clear();
+        tgt->dst.clear();
+        tgt->src.id = data_to_node_id(msg + MSG_SRC_OFS);
+        // now contains the length of the message after the header.
+        int payload_bytes = expected_size - HDR_LEN;
+        int payload_ofs;
+        if (Defs::get_mti_address(tgt->mti))
+        {
+            payload_ofs = MSG_ADR_PAYLOAD_OFS;
+            tgt->dst.id = data_to_node_id(msg + MSG_DST_OFS);
+        }
+        else
+        {
+            payload_ofs = MSG_GLOBAL_PAYLOAD_OFS;
+        }
+        payload_bytes -= payload_ofs;
+        if (payload_bytes < 0)
+        {
+            LOG(WARNING, "TCP message to short.");
+            return false;
+        }
+        tgt->payload.assign(msg + payload_ofs, payload_bytes);
+        return true;
+    }
+
+    /// @param tcp_payload is a message holding a TCP protocol frame.
+    /// @return the OpenLCB priority in range 0..3 or uint_max if the message
+    /// is broken.
+    static unsigned guess_priority(const string &tcp_payload)
+    {
+        if (tcp_payload.size() < ABS_MTI_OFS + 2)
+        {
+            return UINT_MAX;
+        }
+        auto mti = (Defs::MTI)data_to_error(tcp_payload.data() + ABS_MTI_OFS);
+        return Defs::mti_priority(mti);
+    }
+
+    enum
```

Comment:
Consider commenting the members of this enum

Reply:
done